x/harvest/keeper/borrow.go

```diff
@@ -0,0 +1,138 @@
+package keeper
+
+import (
+	"math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+
+	"github.com/kava-labs/kava/x/harvest/types"
+)
+
+// USDX is the USDX coin's denom
+const USDX = "usdx"
+
+// Borrow funds
+func (k Keeper) Borrow(ctx sdk.Context, borrower sdk.AccAddress, amount sdk.Coin) error {
+
+	err := k.ValidateBorrow(ctx, borrower, amount)
+	if err != nil {
+		return err
+	}
+
+	err = k.supplyKeeper.SendCoinsFromAccountToModule(ctx, borrower, types.ModuleAccountName, sdk.NewCoins(amount))
+	if err != nil {
+		return err
+	}
+
+	borrow, found := k.GetBorrow(ctx, borrower, amount.Denom)
+	if !found {
+		borrow = types.NewBorrow(borrower, amount)
+	} else {
+		borrow.Amount = borrow.Amount.Add(amount)
+	}
+
+	k.SetBorrow(ctx, borrow)
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeHarvestBorrow,
+			sdk.NewAttribute(sdk.AttributeKeyAmount, amount.String()),
+			sdk.NewAttribute(types.AttributeKeyBorrower, borrow.Borrower.String()),
+			sdk.NewAttribute(types.AttributeKeyBorrowDenom, borrow.Amount.Denom),
+		),
+	)
+
+	return nil
+}
+
+// ValidateBorrow validates a borrow request against borrower and protocol requirements
+func (k Keeper) ValidateBorrow(ctx sdk.Context, borrower sdk.AccAddress, amount sdk.Coin) error {
+	if err := k.validateBorrowUser(ctx, borrower, amount); err != nil {
+		return err
+	}
+
+	if err := k.validateBorrowSystem(ctx, borrower, amount); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (k Keeper) validateBorrowUser(ctx sdk.Context, borrower sdk.AccAddress, amount sdk.Coin) error {
+	var borrowMoneyMarket types.MoneyMarket
+	var borrowAssetPrice sdk.Dec
+	var borrowUSDValue sdk.Dec
+	if amount.Denom == USDX {
+		borrowUSDValue = sdk.NewDecFromInt(amount.Amount).Quo(sdk.NewDec(int64(math.Pow(10, 6)))) // TODO: include conversion_factor in params
```

Comment:
Resolve this todo before merge - code is too hard to follow with these USDX specific if/else statements

Reply:
Will add `conversion_factor`, which will replace the `math.Pow(10, 6)` here. But the specific USDX if/else condition is to avoid checking the pricefeed for the USDX price when it would return an error as there's no USDX market - so we'll need to add an additional parameter `stablecoin` in order to replace the USDX specific if/else condition.

In this proposal the if/else would remain, but would check for `param.Stablecoin == true` instead of `amount.Denom == "usdx"`. This solution doesn't feel great since BUSD is also a stablecoin but exists in the pricefeed so we can check for price the normal way.

Reply:
Unless we'll add a pricefeed for USDX? I'm thinking that we shouldn't do this as the protocol naturally assumes that USDX's price is $1.00.

Reply:
Addressed the `conversation_factor` part of this in https://github.com/Kava-Labs/kava/pull/702/commits/277c4b1936981182fc09ae0c6ea0973d9cbbca8a.

Reply:
We'll add a pricefeed for USDX